File: backend/app/api/tasks.py
from fastapi import APIRouter, status
from sqlalchemy import func, select

from app.api.members import get_member
from app.auth import DbSession, ParentSession
from app.errors import ApiError
from app.models import Task, TaskAssignment, TaskRecurrence
from app.schemas import MemberPositionOut, TaskIn, TaskOrderIn, TaskOut

router = APIRouter(tags=["tasks"])


def task_out(task: Task) -> TaskOut:
    recurrence = task.recurrence
    return TaskOut(
        id=task.id,
        title=task.title,
        icon=task.icon,
        description=task.description or "",
        points=task.points,
        time_of_day=task.time_of_day,
        color=task.color,
        active=task.active,
        needs_approval=task.needs_approval,
        shared=task.shared,
        extra=task.extra,
        recurrence={
            "kind": recurrence.kind,
            "weekdays": recurrence.weekdays,
            "date": recurrence.date,
            "interval_days": recurrence.interval_days,
        },
        member_ids=sorted(assignment.member_id for assignment in task.assignments),
        positions=positions_out(task),
    )


def positions_out(task: Task) -> list[MemberPositionOut]:
    return [
        MemberPositionOut(member_id=a.member_id, position=a.position)
        for a in sorted(task.assignments, key=lambda a: a.member_id)
    ]


def get_task(db: DbSession, task_id: int) -> Task:
    task = db.get(Task, task_id)
    if task is None:
        raise ApiError(status.HTTP_404_NOT_FOUND, "task.not_found")
    return task


def apply_task(db: DbSession, task: Task, body: TaskIn) -> None:
    for member_id in body.member_ids:
        get_member(db, member_id)

    task.title, task.icon, task.points = body.title, body.icon, body.points
    task.description = body.description or None
    task.time_of_day, task.color, task.active = body.time_of_day, body.color, body.active
    task.needs_approval, task.shared, task.extra = body.needs_approval, body.shared, body.extra

    rule = body.recurrence
    if task.recurrence is None:
        task.recurrence = TaskRecurrence()
    task.recurrence.kind = rule.kind
    task.recurrence.weekdays = getattr(rule, "weekdays", None)
    task.recurrence.date = getattr(rule, "date", None)
    task.recurrence.interval_days = getattr(rule, "interval_days", None)

    # Bestehende Zuordnungen behalten, damit sie ihre Id und ihren Platz nicht wechseln;
    # neue Personen bekommen die Aufgabe ans Ende ihrer Reihenfolge.
    wanted = set(body.member_ids)
    task.assignments = [a for a in task.assignments if a.member_id in wanted] + [
        TaskAssignment(member_id=member_id, position=_next_position(db, member_id))
        for member_id in sorted(wanted - {a.member_id for a in task.assignments})
    ]


def _next_position(db: DbSession, member_id: int) -> int:
    last = db.scalar(
        select(func.max(TaskAssignment.position)).where(TaskAssignment.member_id == member_id)
    )
    return 0 if last is None else last + 1


@router.get("/tasks")
def list_tasks(_: ParentSession, db: DbSession) -> list[TaskOut]:
    tasks = db.scalars(select(Task).order_by(Task.id))
    return [task_out(task) for task in tasks]


@router.post("/tasks", status_code=status.HTTP_201_CREATED)
def create_task(body: TaskIn, _: ParentSession, db: DbSession) -> TaskOut:
    task = Task()
    apply_task(db, task, body)
    db.add(task)
    db.commit()
    return task_out(task)


@router.put("/tasks/{task_id}")
def update_task(task_id: int, body: TaskIn, _: ParentSession, db: DbSession) -> TaskOut:
    task = get_task(db, task_id)
    apply_task(db, task, body)
    db.commit()
    return task_out(task)


@router.delete("/tasks/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task(task_id: int, _: ParentSession, db: DbSession) -> None:
    db.delete(get_task(db, task_id))
    db.commit()


@router.put("/members/{member_id}/task-order", status_code=status.HTTP_204_NO_CONTENT)
def set_task_order(member_id: int, body: TaskOrderIn, _: ParentSession, db: DbSession) -> None:
    """Reihenfolge der Aufgaben einer Person, z. B. die Schritte ihrer Morgenroutine.

    Genannte Aufgaben kommen in dieser Reihenfolge nach vorn, alle anderen behalten ihre
    bisherige Abfolge dahinter.
    """
    get_member(db, member_id)
    assignments = db.scalars(
        select(TaskAssignment)
        .where(TaskAssignment.member_id == member_id)
        .order_by(TaskAssignment.position, TaskAssignment.task_id)
    ).all()
    by_task = {a.task_id: a for a in assignments}
    if len(set(body.task_ids)) != len(body.task_ids) or any(
        task_id not in by_task for task_id in body.task_ids
    ):
        raise ApiError(status.HTTP_409_CONFLICT, "task.order_invalid")
    listed = [by_task[task_id] for task_id in body.task_ids]
    rest = [a for a in assignments if a.task_id not in set(body.task_ids)]
    for position, assignment in enumerate(listed + rest):
        assignment.position = position
    db.commit()

File: frontend/src/api/tasks.ts
import { useQuery, useQueryClient } from '@tanstack/react-query'

import type { MemberColor } from '../memberColors'
import { api, apiGet } from './client'
import { useParentMutation } from './mutations'
import { TODAY_KEY } from './today'

// Müssen zu TIMES_OF_DAY bzw. den Recurrence-Modellen im Backend passen (backend/app/schemas.py).
export const TIMES_OF_DAY = ['morning', 'midday', 'afternoon', 'evening'] as const
export type TimeOfDay = (typeof TIMES_OF_DAY)[number]
export const TASK_MAX_POINTS = 1000
export const MAX_INTERVAL_DAYS = 365

/** Wochentage nach ISO: 1 = Montag … 7 = Sonntag. Datum als `YYYY-MM-DD`. */
export type Recurrence =
  | { kind: 'daily' }
  | { kind: 'weekly'; weekdays: number[] }
  | { kind: 'once'; date: string }
  /** Ohne festen Tag: fällig ab `date`, danach `interval_days` nach der letzten Erledigung. */
  | { kind: 'flexible'; interval_days: number; date: string }
export type RecurrenceKind = Recurrence['kind']

export interface TaskData {
  title: string
  icon: string
  description: string
  points: number
  time_of_day: TimeOfDay | null
  /** null = Farbe der jeweiligen Person */
  color: MemberColor | null
  active: boolean
  /** Punkte erst nach Kontrolle durch die Eltern. */
  needs_approval: boolean
  /** „Einer für alle“: eine Erledigung gilt für alle zugeordneten Personen. */
  shared: boolean
  /** Extra-Aufgabe: freiwillig, außerhalb der Routinen, zählt nicht zum Tagesfortschritt. */
  extra: boolean
  recurrence: Recurrence
  member_ids: number[]
}

/** Platz einer Aufgabe in der Reihenfolge einer Person (kleiner = früher). */
export interface MemberPosition {
  member_id: number
  position: number
}

export interface Task extends TaskData {
  id: number
  positions: MemberPosition[]
}

/** Block einer Aufgabe: Routine eines Tagesabschnitts, „Jederzeit“ (null) oder Extra. */
export type TaskBlock = TimeOfDay | null | 'extra'
export const TASK_BLOCKS: readonly TaskBlock[] = [...TIMES_OF_DAY, null, 'extra']

interface Orderable {
  id: number
  time_of_day: TimeOfDay | null
  extra: boolean
  positions: MemberPosition[]
}

export const blockOf = (task: Orderable): TaskBlock => (task.extra ? 'extra' : task.time_of_day)

/**
 * Aufgaben einer Person in ihrer Reihenfolge am Display: Blöcke nacheinander (Tagesabschnitte,
 * „Jederzeit“, Extras), innerhalb eines Blocks in der von den Eltern festgelegten Abfolge.
 */
export function sortForMember<T extends Orderable>(tasks: T[], memberId: number): T[] {
  const block = (task: T) => TASK_BLOCKS.indexOf(blockOf(task))
  return [...tasks].sort(
    (a, b) =>
      block(a) - block(b) || positionFor(a, memberId) - positionFor(b, memberId) || a.id - b.id,
  )
}

/** Platz in der Reihenfolge dieser Person; ohne Eintrag ans Ende. */
export function positionFor(task: { positions: MemberPosition[] }, memberId: number) {
  return (
    task.positions.find((entry) => entry.member_id === memberId)?.position ??
    Number.MAX_SAFE_INTEGER
  )
}

export const TASKS_KEY = ['tasks'] as const

export function useTasks() {
  return useQuery({ queryKey: TASKS_KEY, queryFn: () => apiGet<Task[]>('/tasks') })
}

export const createTask = (data: TaskData) => api<Task>('POST', '/tasks', data)
export const updateTask = (id: number, data: TaskData) => api<Task>('PUT', `/tasks/${id}`, data)
export const deleteTask = (id: number) => api<void>('DELETE', `/tasks/${id}`)

export function taskData({ id: _id, positions: _positions, ...data }: Task): TaskData {
  return data
}

interface TaskOrder {
  memberId: number
  /** Alle Aufgaben der Person in der neuen Reihenfolge. */
  taskIds: number[]
}

const withOrder = (tasks: Task[], { memberId, taskIds }: TaskOrder) =>
  tasks.map((task) => {
    const position = taskIds.indexOf(task.id)
    if (position < 0) return task
    return {
      ...task,
      positions: task.positions.map((entry) =>
        entry.member_id === memberId ? { ...entry, position } : entry,
      ),
    }
  })

/** Reihenfolge der Aufgaben einer Person (z. B. Schritte der Morgenroutine); sofort sichtbar. */
export function useSetTaskOrder() {
  const queryClient = useQueryClient()
  const mutation = useTasksMutation((order: TaskOrder) =>
    api<void>('PUT', `/members/${order.memberId}/task-order`, { task_ids: order.taskIds }),
  )
  return {
    ...mutation,
    mutate: (order: TaskOrder) => {
      queryClient.setQueryData<Task[]>(TASKS_KEY, (tasks) => tasks && withOrder(tasks, order))
      mutation.mutate(order)
    },
  }
}

export function useTasksMutation<TVariables, TResult>(
  request: (variables: TVariables) => Promise<TResult>,
) {
  return useParentMutation(request, [TASKS_KEY, TODAY_KEY])
}
